File: sem4/mpp/project-cs/networking/Worker/MasterJsonWorker.cs
using System.Net.Sockets;
using project_cs.domain;
using project_cs.utils;

namespace networking.Worker;

public class MasterJsonWorker : AbstractJsonWorker, IUserObserver {
    private IMasterService Service;

    public MasterJsonWorker(IMasterService service, TcpClient connection) : base(connection) {
        Service = service;
    }

    protected override Response HandleRequest(Request request) {
        Response response = null;
        Console.WriteLine(request);
        switch (request.type) {
            case RequestType.LOGIN: {
                Console.WriteLine("Login request ...");
                var user = request.user;
                try {
                    Service.Login(user.username, user.hashedPassword, this);
                    return okResponse;
                }
                catch (Exception e) {
//				connected = false;
                    response = new Response {
                        type = ResponseType.ERROR,
                        errorMessage = e.Message
                    };
                    return response;
                }
            }
            case RequestType.REGISTER: {
                Console.WriteLine("Register request ...");
                var user = request.user;
                try {
                    Service.Register(user.username, user.hashedPassword);
                    return okResponse;
                }
                catch (Exception e) {
//				connected = false;
                    response = new Response {
                        type = ResponseType.ERROR,
                        errorMessage = e.Message
                    };
                    return response;
                }
            }
            case RequestType.GET_CHILDREN: {
                Console.WriteLine("Children request ...");
                var search = request.search;
                try {
                    return new Response {
                        type = ResponseType.CHILDREN_LIST,
                        children = Service.SearchChildren(search.raceDistance, search.ageGroup)
                    };
                }
                catch (Exception e) {
//				connected = false;
                    response = new Response {
                        type = ResponseType.ERROR,
                        errorMessage = e.Message
                    };
                    return response;
                }
            }
            case RequestType.SIGN_UP_CHILD: {
                Console.WriteLine("Sign up");
                var signUp = request.signup;
                try {
                    Service.SignUpChild(signUp.name, signUp.age, signUp.raceDistance, signUp.ageGroup);
                    return okResponse;
                }
                catch (Exception e) {
//				connected = false;
                    response = new Response {
                        type = ResponseType.ERROR,
                        errorMessage = e.Message
                    };
                    return response;
                }
            }
            case RequestType.LOGOUT:
                connected = false;
                Service.Logout(this);
                break;
        }

        return response;
    }

    public void NotifySignup() {
        var response = new Response {
            type = ResponseType.UPDATE_CHILDREN
        };
        Console.WriteLine("Received signup notification");
        try {
            SendResponse(response);
        }
        catch (Exception e) {
            Console.WriteLine(e.StackTrace);
        }
    }
}

File: sem4/mpp/project-java/networking/src/main/java/networking/worker/MasterJsonWorker.java
package networking.worker;

import domain.User;
import domain.dto.SearchDTO;
import domain.dto.SignUpDTO;
import networking.Request;
import networking.RequestType;
import networking.Response;
import networking.ResponseType;
import services.IMasterService;
import services.IUserObserver;

import java.net.Socket;
import javax.jms.*;


public class MasterJsonWorker extends AbstractJsonWorker implements IUserObserver {
	protected final IMasterService service;


	public MasterJsonWorker(IMasterService service, Socket connection) {
		super(connection);
		this.service = service;
	}

	@Override
	protected Response handleRequest(Request request) {
		Response response = null;
		System.out.println(request);
		if (request.getType() == RequestType.LOGIN) {
			System.out.println("Login request ..." + request.getType());
			User user = request.getUser();
			try {
				service.logIn(user.getUsername(), user.getHashedPassword(), this);
				return okResponse;
			} catch (Exception e) {
//				connected = false;
				response = new Response();
				response.setType(ResponseType.ERROR);
				response.setErrorMessage(e.getMessage());
				return response;
			}
		} else if (request.getType() == RequestType.REGISTER) {
			System.out.println("Register request ..." + request.getType());
			User user = request.getUser();
			try {
				service.register(user.getUsername(), user.getHashedPassword());
				return okResponse;
			} catch (Exception e) {
//				connected = false;
				response = new Response();
				response.setType(ResponseType.ERROR);
				response.setErrorMessage(e.getMessage());
				return response;
			}
		} else if (request.getType() == RequestType.GET_CHILDREN) {
			System.out.println("Getting children ..." + request.getType());
			SearchDTO search = request.getSearch();
			try {
				response = new Response();
				response.setType(ResponseType.CHILDREN_LIST);
				response.setChildren(service.searchChildren(search.getRaceDistance(), search.getAgeGroup()));
				return response;
			} catch (Exception e) {
//				connected = false;
				response = new Response();
				response.setType(ResponseType.ERROR);
				response.setErrorMessage(e.getMessage());
				return response;
			}
		} else if (request.getType() == RequestType.SIGN_UP_CHILD) {
			System.out.println("Getting children ..." + request.getType());
			SignUpDTO signUp = request.getSignup();
			try {
				service.signUpChild(signUp.getName(), signUp.getAge(), signUp.getRaceDistance(), signUp.getAgeGroup());
				return okResponse;
			} catch (Exception e) {
//				connected = false;
				response = new Response();
				response.setType(ResponseType.ERROR);
				response.setErrorMessage(e.getMessage());
				return response;
			}
		} else if (request.getType() == RequestType.LOGOUT) {
			connected = false;
			try {
				service.logOut(this);
			} catch (Exception ignored) {
			}
		}
		return response;
	}

	//	@Override
//	public void notifySignup() {
//		Response response = new Response();
//		response.setType(ResponseType.UPDATE_CHILDREN);
//		System.out.println("Received signup notification");
//		try {
//			sendResponse(response);
//		} catch (Exception e) {
//			e.printStackTrace();
//		}
//	}

	@Override
	public void notifySignup() {
	}

}
